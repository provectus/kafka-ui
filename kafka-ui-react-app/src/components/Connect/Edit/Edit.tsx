import React from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { Controller, useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import { yupResolver } from '@hookform/resolvers/yup';
import { Connector } from 'generated-sources';
import {
  ClusterName,
  ConnectName,
  ConnectorConfig,
  ConnectorName,
} from 'redux/interfaces';
import { clusterConnectConnectorConfigPath } from 'lib/paths';
import yup from 'lib/yupExtended';
import JSONEditor from 'components/common/JSONEditor/JSONEditor';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { Button } from 'components/common/Button/Button';

import {
  ConnectEditWarningMessageStyled,
  ConnectEditWrapperStyled,
} from './Edit.styled';

const validationSchema = yup.object().shape({
  config: yup.string().required().isJsonObject(),
});

interface RouterParams {
  clusterName: ClusterName;
  connectName: ConnectName;
  connectorName: ConnectorName;
}

interface FormValues {
  config: string;
}

export interface EditProps {
  fetchConfig(
    clusterName: ClusterName,
    connectName: ConnectName,
    connectorName: ConnectorName
  ): Promise<void>;
  isConfigFetching: boolean;
  config: ConnectorConfig | null;
  updateConfig(
    clusterName: ClusterName,
    connectName: ConnectName,
    connectorName: ConnectorName,
    connectorConfig: ConnectorConfig
  ): Promise<Connector | undefined>;
}

const Edit: React.FC<EditProps> = ({
  fetchConfig,
  isConfigFetching,
  config,
  updateConfig,
}) => {
  const { clusterName, connectName, connectorName } = useParams<RouterParams>();
  const history = useHistory();
  const {
    handleSubmit,
    control,
    formState: { isDirty, isSubmitting, isValid, errors },
    setValue,
  } = useForm<FormValues>({
    mode: 'onTouched',
    resolver: yupResolver(validationSchema),
    defaultValues: {
      config: JSON.stringify(config, null, '\t'),
    },
  });

  React.useEffect(() => {
    fetchConfig(clusterName, connectName, connectorName);
  }, [fetchConfig, clusterName, connectName, connectorName]);

  React.useEffect(() => {
    if (config) {
      setValue('config', JSON.stringify(config, null, '\t'));
    }
  }, [config, setValue]);

  const onSubmit = React.useCallback(
    async (values: FormValues) => {
      const connector = await updateConfig(
        clusterName,
        connectName,
        connectorName,
        JSON.parse(values.config.trim())
      );
      if (connector) {
        history.push(
          clusterConnectConnectorConfigPath(
            clusterName,
            connectName,
            connectorName
          )
        );
      }
    },
    [updateConfig, clusterName, connectName, connectorName]
  );

  if (isConfigFetching) return <PageLoader />;

  const hasCredentials = JSON.stringify(config, null, '\t').includes(
    '"******"'
  );
  return (
    <ConnectEditWrapperStyled>
      {hasCredentials && (
        <ConnectEditWarningMessageStyled>
          Please replace ****** with the real credential values to avoid
          accidentally breaking your connector config!
        </ConnectEditWarningMessageStyled>
      )}
      <form onSubmit={handleSubmit(onSubmit)} aria-label="Edit connect form">
        <div>
          <Controller
            control={control}
            name="config"
            render={({ field }) => (
              <JSONEditor {...field} readOnly={isSubmitting} />
            )}
          />
        </div>
        <div>
          <ErrorMessage errors={errors} name="config" />
        </div>
        <Button
          buttonSize="M"
          buttonType="primary"
          type="submit"
          disabled={!isValid || isSubmitting || !isDirty}
        >
          Submit
        </Button>
      </form>
    </ConnectEditWrapperStyled>
  );
};

export default Edit;
