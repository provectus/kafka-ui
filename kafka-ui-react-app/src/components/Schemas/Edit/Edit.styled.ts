import styled, { css } from 'styled-components';

export const EditWrapper = styled.div`
  padding: 16px;
  padding-top: 0;
  & > form {
    display: flex;
    flex-direction: column;
    gap: 16px;

    & > div:first-child {
      display: flex;
      gap: 16px;

      & > * {
        width: 20%;
      }
    }

    & > button:last-child {
      width: 72px;
      align-self: center;
    }
  }
`;

export const EditorsWrapper = styled.div`
  display: flex;
  gap: 16px;

  & > * {
    flex-grow: 1;
  }
`;

export const EditorContainer = styled.div(
  ({ theme }) => css`
    border: 1px solid ${theme.layout.stuffBorderColor};
    border-radius: 8px;
    margin-bottom: 16px;
    padding: 16px;
    & > h4 {
      font-weight: 500;
      font-size: 16px;
      line-height: 24px;
      padding-bottom: 16px;
      color: ${theme.heading.h4};
    }
  `
);
