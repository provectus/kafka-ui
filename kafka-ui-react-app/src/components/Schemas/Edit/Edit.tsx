import Breadcrumb from 'components/common/Breadcrumb/Breadcrumb';
import {
  CompatibilityLevelCompatibilityEnum,
  SchemaSubject,
  SchemaType,
} from 'generated-sources';
import { clusterSchemaPath, clusterSchemasPath } from 'lib/paths';
import React from 'react';
import { ClusterName, NewSchemaSubjectRaw, SchemaName } from 'redux/interfaces';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { useHistory } from 'react-router';
import JSONEditor from 'components/common/JSONEditor/JSONEditor';
import { useForm } from 'react-hook-form';

export interface EditProps {
  subject: SchemaName;
  schema: SchemaSubject;
  clusterName: ClusterName;
  schemasAreFetched: boolean;
  fetchSchemasByClusterName: (clusterName: ClusterName) => void;
  updateSchema: (
    latestSchema: SchemaSubject,
    newSchema: string,
    newSchemaType: SchemaType,
    newCompatibilityLevel: CompatibilityLevelCompatibilityEnum,
    clusterName: string,
    subject: string
  ) => Promise<void>;
}

const Edit = ({
  subject,
  schema,
  clusterName,
  schemasAreFetched,
  fetchSchemasByClusterName,
  updateSchema,
}: EditProps) => {
  React.useEffect(() => {
    if (!schemasAreFetched) fetchSchemasByClusterName(clusterName);
  }, [clusterName, fetchSchemasByClusterName]);

  const {
    register,
    handleSubmit,
    formState: { isSubmitting },
    control,
  } = useForm<NewSchemaSubjectRaw>({ mode: 'onChange' });

  const getFormattedSchema = React.useCallback(
    () => JSON.stringify(JSON.parse(schema.schema), null, '\t'),
    [schema]
  );
  const history = useHistory();
  const onSubmit = React.useCallback(
    async ({
      schemaType,
      compatibilityLevel,
      newSchema,
    }: {
      schemaType: SchemaType;
      compatibilityLevel: CompatibilityLevelCompatibilityEnum;
      newSchema: string;
    }) => {
      await updateSchema(
        schema,
        newSchema,
        schemaType,
        compatibilityLevel,
        clusterName,
        subject
      );
      history.push(clusterSchemaPath(clusterName, subject));
    },
    [schema, register, control, clusterName, subject, updateSchema, history]
  );

  return (
    <div className="section">
      <div className="level">
        <div className="level-item level-left">
          <Breadcrumb
            links={[
              {
                href: clusterSchemasPath(clusterName),
                label: 'Schema Registry',
              },
              {
                href: clusterSchemaPath(clusterName, subject),
                label: subject,
              },
            ]}
          >
            Edit
          </Breadcrumb>
        </div>
      </div>

      {schemasAreFetched && !isSubmitting ? (
        <div className="box">
          <form
            onSubmit={handleSubmit(onSubmit)}
            className="mt-3 is-flex-direction-column"
          >
            <div className="mb-4">
              <h5 className="title is-5 mb-2">Schema Type</h5>
              <div className="select">
                <select
                  name="schemaType"
                  ref={register({
                    required: 'Schema Type is required.',
                  })}
                  defaultValue={schema.schemaType}
                >
                  {Object.keys(SchemaType).map((type: string) => (
                    <option key={type} value={type}>
                      {type}
                    </option>
                  ))}
                </select>
              </div>
            </div>

            <div className="mb-4">
              <h5 className="title is-5 mb-2">Compatibility Level</h5>
              <div className="select">
                <select
                  name="compatibilityLevel"
                  ref={register()}
                  defaultValue={schema.compatibilityLevel}
                >
                  {Object.keys(CompatibilityLevelCompatibilityEnum).map(
                    (level: string) => (
                      <option key={level} value={level}>
                        {level}
                      </option>
                    )
                  )}
                </select>
              </div>
            </div>
            <div className="columns">
              <div className="column is-one-half">
                <h4 className="title is-5 mb-2">Latest Schema</h4>
                <JSONEditor
                  readonly
                  value={getFormattedSchema()}
                  name="latestSchema"
                />
              </div>
              <div className="column is-one-half">
                <h4 className="title is-5 mb-2">New Schema</h4>
                <JSONEditor
                  control={control}
                  value={getFormattedSchema()}
                  name="newSchema"
                />
              </div>
            </div>
            <button type="submit" className="button is-primary">
              Submit
            </button>
          </form>
        </div>
      ) : (
        <PageLoader />
      )}
    </div>
  );
};

export default Edit;
