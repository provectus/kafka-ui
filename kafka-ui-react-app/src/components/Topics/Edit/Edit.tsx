import React from 'react';
import {
  ClusterName,
  TopicFormData,
  TopicName,
  TopicConfigByName,
  TopicWithDetailedInfo,
  CleanupPolicy,
} from 'redux/interfaces';
import { useForm, FormContext } from 'react-hook-form';
import { camelCase } from 'lodash';

import TopicForm from '../shared/Form/TopicForm';
import FormBreadcrumbs from '../shared/Form/FormBreadcrumbs';

interface Props {
  clusterName: ClusterName;
  topicName: TopicName;
  topic?: TopicWithDetailedInfo;
  isFetched: boolean;
  isTopicDetailsFetched: boolean;
  isTopicUpdated: boolean;
  fetchTopicDetails: (clusterName: ClusterName, topicName: TopicName) => void;
  fetchTopicConfig: (clusterName: ClusterName, topicName: TopicName) => void;
  updateTopic: (clusterName: ClusterName, form: TopicFormData) => void;
  redirectToTopicPath: (clusterName: ClusterName, topicName: TopicName) => void;
  resetUploadedState: () => void;
}

const DEFAULTS = {
  partitions: 1,
  replicationFactor: 1,
  minInSyncReplicas: 1,
  cleanupPolicy: CleanupPolicy.Delete,
  retentionBytes: -1,
  maxMessageBytes: 1000012,
};

const topicParams = (topic: TopicWithDetailedInfo | undefined) => {
  if (!topic) {
    return DEFAULTS;
  }

  const { name, replicationFactor } = topic;

  const configs = topic.config?.reduce(
    (result: { [name: string]: string }, param) => {
      result[camelCase(param.name)] = param.value || param.defaultValue;
      return result;
    },
    {}
  );

  return {
    ...DEFAULTS,
    name,
    partitions: topic.partitionCount || DEFAULTS.partitions,
    replicationFactor,
    ...configs,
  };
};

let formInit = false;

const Edit: React.FC<Props> = ({
  clusterName,
  topicName,
  topic,
  isFetched,
  isTopicDetailsFetched,
  isTopicUpdated,
  fetchTopicDetails,
  fetchTopicConfig,
  updateTopic,
  redirectToTopicPath,
}) => {
  const defaultValues = topicParams(topic);

  const methods = useForm<TopicFormData>({ defaultValues });

  const [isSubmitting, setIsSubmitting] = React.useState<boolean>(false);

  React.useEffect(() => {
    fetchTopicConfig(clusterName, topicName);
    fetchTopicDetails(clusterName, topicName);
  }, [fetchTopicConfig, fetchTopicDetails, clusterName, topicName]);

  React.useEffect(() => {
    if (isSubmitting && isTopicUpdated) {
      const { name } = methods.getValues();
      redirectToTopicPath(clusterName, name);
    }
  }, [isSubmitting, isTopicUpdated, redirectToTopicPath, clusterName, methods]);

  if (!isFetched || !isTopicDetailsFetched || !topic || !topic.config) {
    return null;
  }

  if (!formInit) {
    methods.reset(defaultValues);
    formInit = true;
  }

  const config: TopicConfigByName = {
    byName: {},
  };

  topic.config.forEach((param) => {
    config.byName[param.name] = param;
  });

  const onSubmit = async (data: TopicFormData) => {
    setIsSubmitting(true);
    updateTopic(clusterName, data);
  };

  return (
    <div className="section">
      <div className="level">
        <FormBreadcrumbs
          clusterName={clusterName}
          topicName={topicName}
          current="Edit Topic"
        />
      </div>

      <div className="box">
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <FormContext {...methods}>
          <TopicForm
            topicName={topicName}
            config={config}
            isSubmitting={isSubmitting}
            isEditing
            onSubmit={methods.handleSubmit(onSubmit)}
          />
        </FormContext>
      </div>
    </div>
  );
};

export default Edit;
