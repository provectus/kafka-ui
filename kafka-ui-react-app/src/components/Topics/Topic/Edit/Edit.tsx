import React from 'react';
import { TopicConfigByName, TopicFormData } from 'redux/interfaces';
import { useForm, FormProvider } from 'react-hook-form';
import TopicForm from 'components/Topics/shared/Form/TopicForm';
import { RouteParamsClusterTopic } from 'lib/paths';
import { useNavigate } from 'react-router-dom';
import { yupResolver } from '@hookform/resolvers/yup';
import { topicFormValidationSchema } from 'lib/yupExtended';
import useAppParams from 'lib/hooks/useAppParams';
import topicParamsTransformer from 'components/Topics/Topic/Edit/topicParamsTransformer';
import { MILLISECONDS_IN_WEEK } from 'lib/constants';
import {
  useTopicConfig,
  useTopicDetails,
  useUpdateTopic,
} from 'lib/hooks/api/topics';
import DangerZone from 'components/Topics/Topic/Edit/DangerZone/DangerZone';
import { ConfigSource } from 'generated-sources';

export const TOPIC_EDIT_FORM_DEFAULT_PROPS = {
  partitions: 1,
  replicationFactor: 1,
  minInSyncReplicas: 1,
  cleanupPolicy: 'delete',
  retentionBytes: -1,
  retentionMs: MILLISECONDS_IN_WEEK,
  maxMessageBytes: 1000012,
  customParams: [],
};

const Edit: React.FC = () => {
  const { clusterName, topicName } = useAppParams<RouteParamsClusterTopic>();
  const { data: topic } = useTopicDetails({ clusterName, topicName });
  const { data: topicConfig } = useTopicConfig({ clusterName, topicName });
  const updateTopic = useUpdateTopic({ clusterName, topicName });

  const defaultValues = topicParamsTransformer(topic, topicConfig);

  const methods = useForm<TopicFormData>({
    defaultValues,
    resolver: yupResolver(topicFormValidationSchema),
    mode: 'onChange',
  });

  const navigate = useNavigate();

  const config: TopicConfigByName = {
    byName: {},
  };

  topicConfig?.forEach((param) => {
    config.byName[param.name] = param;
  });
  const onSubmit = async (data: TopicFormData) => {
    const filteredDirtyDefaultEntries = Object.entries(data).filter(
      ([key, val]) => {
        const isDirty =
          String(val) !==
          String(defaultValues[key as keyof typeof defaultValues]);

        const isDefaultConfig =
          config.byName[key]?.source === ConfigSource.DEFAULT_CONFIG;

        // if it is changed should be sent or if it was Dynamic
        return isDirty || !isDefaultConfig;
      }
    );

    const newData = Object.fromEntries(filteredDirtyDefaultEntries);
    try {
      await updateTopic.mutateAsync(newData);
      navigate('../');
    } catch (e) {
      // do nothing
    }
  };

  return (
    <>
      <FormProvider {...methods}>
        <TopicForm
          config={config.byName}
          topicName={topicName}
          retentionBytes={defaultValues.retentionBytes}
          inSyncReplicas={Number(defaultValues.minInSyncReplicas)}
          isSubmitting={updateTopic.isLoading}
          cleanUpPolicy={topic?.cleanUpPolicy}
          isEditing
          onSubmit={methods.handleSubmit(onSubmit)}
        />
      </FormProvider>
      {topic && (
        <DangerZone
          defaultPartitions={defaultValues.partitions}
          defaultReplicationFactor={
            defaultValues.replicationFactor ||
            TOPIC_EDIT_FORM_DEFAULT_PROPS.replicationFactor
          }
        />
      )}
    </>
  );
};

export default Edit;
