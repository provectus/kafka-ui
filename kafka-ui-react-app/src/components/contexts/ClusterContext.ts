import React from 'react';

export interface ContextProps {
  isReadOnly: boolean;
  hasKafkaConnectConfigured: boolean;
  hasSchemaRegistryConfigured: boolean;
  isTopicDeletionAllowed: boolean;
}

export const initialValue: ContextProps = {
  isReadOnly: false,
  hasKafkaConnectConfigured: false,
  hasSchemaRegistryConfigured: false,
  isTopicDeletionAllowed: true,
};
const ClusterContext = React.createContext(initialValue);

export default ClusterContext;
