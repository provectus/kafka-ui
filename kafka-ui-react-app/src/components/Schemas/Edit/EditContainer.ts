import { connect } from 'react-redux';
import { ClusterName, RootState } from 'redux/interfaces';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import {
  getIsSchemaListFetched,
  getSchema,
  getSchemaUpdated,
} from 'redux/reducers/schemas/selectors';
import { fetchSchemasByClusterName, updateSchema } from 'redux/actions';

import Edit from './Edit';

interface RouteProps {
  clusterName: ClusterName;
  subject: string;
}

type OwnProps = RouteComponentProps<RouteProps>;

const mapStateToProps = (
  state: RootState,
  {
    match: {
      params: { clusterName, subject },
    },
  }: OwnProps
) => ({
  subject,
  schema: getSchema(state, subject),
  schemasAreFetched: getIsSchemaListFetched(state),
  clusterName,
  isSchemaUpdated: getSchemaUpdated(state),
});

const mapDispatchToProps = {
  fetchSchemasByClusterName,
  updateSchema,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Edit));
