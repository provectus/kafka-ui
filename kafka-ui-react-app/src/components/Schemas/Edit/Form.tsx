import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useForm, Controller, FormProvider } from 'react-hook-form';
import {
  CompatibilityLevelCompatibilityEnum,
  SchemaType,
} from 'generated-sources';
import {
  clusterSchemaPath,
  clusterSchemasPath,
  ClusterSubjectParam,
} from 'lib/paths';
import { NewSchemaSubjectRaw } from 'redux/interfaces';
import Editor from 'components/common/Editor/Editor';
import Select from 'components/common/Select/Select';
import { Button } from 'components/common/Button/Button';
import { InputLabel } from 'components/common/Input/InputLabel.styled';
import PageHeading from 'components/common/PageHeading/PageHeading';
import { useAppDispatch, useAppSelector } from 'lib/hooks/redux';
import useAppParams from 'lib/hooks/useAppParams';
import {
  schemaAdded,
  getSchemaLatest,
  getAreSchemaLatestFulfilled,
  schemaUpdated,
  getAreSchemaLatestRejected,
} from 'redux/reducers/schemas/schemasSlice';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { schemasApiClient } from 'lib/api';
import { showServerError } from 'lib/errorHandling';

import * as S from './Edit.styled';

const Form: React.FC = () => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const { clusterName, subject } = useAppParams<ClusterSubjectParam>();

  const schema = useAppSelector((state) => getSchemaLatest(state));
  const isFetched = useAppSelector(getAreSchemaLatestFulfilled);
  const isRejected = useAppSelector(getAreSchemaLatestRejected);

  const formatedSchema = React.useMemo(() => {
    return schema?.schemaType === SchemaType.PROTOBUF
      ? schema?.schema
      : JSON.stringify(JSON.parse(schema?.schema || '{}'), null, '\t');
  }, [schema]);

  const methods = useForm<NewSchemaSubjectRaw>({
    mode: 'onChange',
    defaultValues: {
      schemaType: schema?.schemaType,
      compatibilityLevel:
        schema?.compatibilityLevel as CompatibilityLevelCompatibilityEnum,
      newSchema: formatedSchema,
    },
  });

  const {
    formState: { isDirty, isSubmitting, dirtyFields },
    control,
    handleSubmit,
  } = methods;

  const onSubmit = async (props: NewSchemaSubjectRaw) => {
    if (!schema) return;

    try {
      if (dirtyFields.compatibilityLevel) {
        await schemasApiClient.updateSchemaCompatibilityLevel({
          clusterName,
          subject,
          compatibilityLevel: {
            compatibility: props.compatibilityLevel,
          },
        });
        dispatch(
          schemaUpdated({
            ...schema,
            compatibilityLevel: props.compatibilityLevel,
          })
        );
      }
      if (dirtyFields.newSchema || dirtyFields.schemaType) {
        const resp = await schemasApiClient.createNewSchema({
          clusterName,
          newSchemaSubject: {
            ...schema,
            schema: props.newSchema || schema.schema,
            schemaType: props.schemaType || schema.schemaType,
          },
        });
        dispatch(schemaAdded(resp));
      }

      navigate(clusterSchemaPath(clusterName, subject));
    } catch (e) {
      showServerError(e as Response);
    }
  };

  if (isRejected) {
    navigate('/404');
  }

  if (!isFetched || !schema) {
    return <PageLoader />;
  }
  return (
    <FormProvider {...methods}>
      <PageHeading
        text={`${subject} Edit`}
        backText="Schema Registry"
        backTo={clusterSchemasPath(clusterName)}
      />
      <S.EditWrapper>
        <form onSubmit={handleSubmit(onSubmit)}>
          <div>
            <div>
              <InputLabel>Type</InputLabel>
              <Controller
                control={control}
                rules={{ required: true }}
                name="schemaType"
                render={({ field: { name, onChange, value } }) => (
                  <Select
                    name={name}
                    value={value}
                    onChange={onChange}
                    minWidth="100%"
                    disabled
                    options={Object.keys(SchemaType).map((type) => ({
                      value: type,
                      label: type,
                    }))}
                  />
                )}
              />
            </div>

            <div>
              <InputLabel>Compatibility level</InputLabel>
              <Controller
                control={control}
                name="compatibilityLevel"
                render={({ field: { name, onChange, value } }) => (
                  <Select
                    name={name}
                    value={value}
                    onChange={onChange}
                    minWidth="100%"
                    disabled={isSubmitting}
                    options={Object.keys(
                      CompatibilityLevelCompatibilityEnum
                    ).map((level) => ({ value: level, label: level }))}
                  />
                )}
              />
            </div>
          </div>
          <S.EditorsWrapper>
            <div>
              <S.EditorContainer>
                <h4>Latest schema</h4>
                <Editor
                  schemaType={schema?.schemaType}
                  isFixedHeight
                  readOnly
                  height="372px"
                  value={formatedSchema}
                  name="latestSchema"
                  highlightActiveLine={false}
                />
              </S.EditorContainer>
            </div>
            <div>
              <S.EditorContainer>
                <h4>New schema</h4>
                <Controller
                  control={control}
                  name="newSchema"
                  render={({ field: { name, onChange, value } }) => (
                    <Editor
                      schemaType={schema?.schemaType}
                      readOnly={isSubmitting}
                      defaultValue={value}
                      name={name}
                      onChange={onChange}
                    />
                  )}
                />
              </S.EditorContainer>
              <Button
                buttonType="primary"
                buttonSize="M"
                type="submit"
                disabled={!isDirty || isSubmitting}
              >
                Submit
              </Button>
            </div>
          </S.EditorsWrapper>
        </form>
      </S.EditWrapper>
    </FormProvider>
  );
};

export default Form;
