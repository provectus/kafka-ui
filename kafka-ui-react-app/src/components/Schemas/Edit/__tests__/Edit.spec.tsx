import React from 'react';
import Edit from 'components/Schemas/Edit/Edit';
import { render, WithRoute } from 'lib/testHelpers';
import { clusterSchemaEditPath } from 'lib/paths';
import {
  schemasInitialState,
  schemaVersion,
  schemaVersionWithNonAsciiChars,
} from 'redux/reducers/schemas/__test__/fixtures';
import { screen } from '@testing-library/dom';
import ClusterContext, {
  ContextProps,
  initialValue as contextInitialValue,
} from 'components/contexts/ClusterContext';
import { RootState } from 'redux/interfaces';
import fetchMock from 'fetch-mock';
import { act } from '@testing-library/react';

const clusterName = 'testClusterName';
const schemasAPILatestUrl = `/api/clusters/${clusterName}/schemas/${schemaVersion.subject}/latest`;

const renderComponent = (
  initialState: RootState['schemas'] = schemasInitialState,
  context: ContextProps = contextInitialValue
) =>
  render(
    <WithRoute path={clusterSchemaEditPath()}>
      <ClusterContext.Provider value={context}>
        <Edit />
      </ClusterContext.Provider>
    </WithRoute>,
    {
      initialEntries: [
        clusterSchemaEditPath(clusterName, schemaVersion.subject),
      ],
      preloadedState: {
        schemas: initialState,
      },
    }
  );

describe('Edit', () => {
  afterEach(() => fetchMock.reset());

  describe('fetch success', () => {
    describe('has schema versions', () => {
      it('renders component with schema info', async () => {
        fetchMock.getOnce(schemasAPILatestUrl, schemaVersion);
        await act(() => {
          renderComponent();
        });
        expect(fetchMock.called(schemasAPILatestUrl)).toBeTruthy();
        expect(screen.getByText('Submit')).toBeInTheDocument();
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });
    });
  });

  describe('fetch success schema with non ascii characters', () => {
    describe('has schema versions', () => {
      it('renders component with schema info', async () => {
        fetchMock.getOnce(schemasAPILatestUrl, schemaVersionWithNonAsciiChars);
        await act(() => {
          renderComponent();
        });
        expect(fetchMock.called(schemasAPILatestUrl)).toBeTruthy();
        expect(screen.getByText('Submit')).toBeInTheDocument();
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });
    });
  });
});
