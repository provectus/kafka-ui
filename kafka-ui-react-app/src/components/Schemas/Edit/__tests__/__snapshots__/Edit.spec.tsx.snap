// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component when schemas are fetched matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/ClusterName/schemas",
              "label": "Schema Registry",
            },
            Object {
              "href": "/ui/clusters/ClusterName/schemas/Subject/latest",
              "label": "Subject",
            },
          ]
        }
      >
        Edit
      </Breadcrumb>
    </div>
  </div>
  <div
    className="box"
  >
    <form
      className="mt-3 is-flex-direction-column"
      onSubmit={[Function]}
    >
      <div
        className="mb-4"
      >
        <h5
          className="title is-5 mb-2"
        >
          Schema Type
        </h5>
        <div
          className="select"
        >
          <select
            defaultValue="AVRO"
            name="schemaType"
          >
            <option
              key="AVRO"
              value="AVRO"
            >
              AVRO
            </option>
            <option
              key="JSON"
              value="JSON"
            >
              JSON
            </option>
            <option
              key="PROTOBUF"
              value="PROTOBUF"
            >
              PROTOBUF
            </option>
          </select>
        </div>
      </div>
      <div
        className="mb-4"
      >
        <h5
          className="title is-5 mb-2"
        >
          Compatibility Level
        </h5>
        <div
          className="select"
        >
          <select
            defaultValue="BACKWARD"
            name="compatibilityLevel"
          >
            <option
              key="BACKWARD"
              value="BACKWARD"
            >
              BACKWARD
            </option>
            <option
              key="BACKWARD_TRANSITIVE"
              value="BACKWARD_TRANSITIVE"
            >
              BACKWARD_TRANSITIVE
            </option>
            <option
              key="FORWARD"
              value="FORWARD"
            >
              FORWARD
            </option>
            <option
              key="FORWARD_TRANSITIVE"
              value="FORWARD_TRANSITIVE"
            >
              FORWARD_TRANSITIVE
            </option>
            <option
              key="FULL"
              value="FULL"
            >
              FULL
            </option>
            <option
              key="FULL_TRANSITIVE"
              value="FULL_TRANSITIVE"
            >
              FULL_TRANSITIVE
            </option>
            <option
              key="NONE"
              value="NONE"
            >
              NONE
            </option>
          </select>
        </div>
      </div>
      <div
        className="columns"
      >
        <div
          className="column is-one-half"
        >
          <h4
            className="title is-5 mb-2"
          >
            Latest Schema
          </h4>
          <JSONEditor
            name="latestSchema"
            readonly={true}
            value="{
	\\"schema\\": \\"schema\\"
}"
          />
        </div>
        <div
          className="column is-one-half"
        >
          <h4
            className="title is-5 mb-2"
          >
            New Schema
          </h4>
          <JSONEditor
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayDefaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayNamesRef": Object {
                  "current": Set {},
                },
                "fieldArrayValuesRef": Object {
                  "current": Object {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "fieldsWithValidationRef": Object {
                  "current": Object {},
                },
                "formState": Object {
                  "dirtyFields": Object {},
                  "errors": Object {},
                  "isDirty": false,
                  "isSubmitSuccessful": false,
                  "isSubmitted": false,
                  "isSubmitting": false,
                  "isValid": true,
                  "isValidating": false,
                  "submitCount": 0,
                  "touched": Object {},
                },
                "formStateRef": Object {
                  "current": Object {
                    "dirtyFields": Object {},
                    "errors": Object {},
                    "isDirty": false,
                    "isSubmitSuccessful": false,
                    "isSubmitted": false,
                    "isSubmitting": false,
                    "isValid": true,
                    "isValidating": false,
                    "submitCount": 0,
                    "touched": Object {},
                  },
                },
                "getValues": [Function],
                "isFormDirty": [Function],
                "mode": Object {
                  "isOnAll": false,
                  "isOnBlur": false,
                  "isOnChange": true,
                  "isOnSubmit": false,
                  "isOnTouch": false,
                },
                "reValidateMode": Object {
                  "isReValidateOnBlur": false,
                  "isReValidateOnChange": true,
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "constructor": true,
                    "dirtyFields": true,
                    "errors": true,
                    "isDirty": true,
                    "isSubmitSuccessful": true,
                    "isSubmitted": true,
                    "isSubmitting": true,
                    "isValid": true,
                    "isValidating": true,
                    "submitCount": true,
                    "touched": true,
                  },
                },
                "register": [Function],
                "removeFieldEventListener": [Function],
                "resetFieldArrayFunctionRef": Object {
                  "current": Object {},
                },
                "setValue": [Function],
                "shallowFieldsStateRef": Object {
                  "current": Object {},
                },
                "shouldUnregister": true,
                "trigger": [Function],
                "unregister": [Function],
                "updateFormState": [Function],
                "updateWatchedValue": [Function],
                "useWatchFieldsRef": Object {
                  "current": Object {},
                },
                "useWatchRenderFunctionsRef": Object {
                  "current": Object {},
                },
                "validFieldsRef": Object {
                  "current": Object {},
                },
                "validateResolver": undefined,
                "watchInternal": [Function],
              }
            }
            name="newSchema"
            value="{
	\\"schema\\": \\"schema\\"
}"
          />
        </div>
      </div>
      <button
        className="button is-primary"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`Edit Component when schemas are not fetched matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/ClusterName/schemas",
              "label": "Schema Registry",
            },
            Object {
              "href": "/ui/clusters/ClusterName/schemas/Subject/latest",
              "label": "Subject",
            },
          ]
        }
      >
        Edit
      </Breadcrumb>
    </div>
  </div>
  <PageLoader />
</div>
`;
